### ISSUE TITLE
Proxy endpoint for services does not appear to be working for client library

### ISSUE THREAD

**Issue by satnam6502** [GoogleCloudPlatform/kubernetes#4977]
I am trying to access the Elasticsearch service via the proxy in our e2e tests:
```
func getURL(c *client.Client, url string) ([]byte, error) {
	body, err := c.Get().Namespace(api.NamespaceDefault).
		AbsPath(fmt.Sprintf("/api/v1beta1/proxy/services/elasticsearch-logging/%s", url)).Do().Raw()
	if err != nil {
		return nil, err
	}
	return body, nil
}

...
	body, err := getURL(c, "")
```
However, this fails although there is mention of a 200 status code:
```
I0302 17:33:14.873064    8431 request.go:508] finalURL = https://104.197.8.236/api/v1beta1/proxy/services/elasticsearch-logging/?namespace=default

• Failure [0.334 seconds]
Cluster level logging using Elasticsearch
/usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:47
  should check that logs from pods on all nodes are ingested into Elasticsearch [It]
  /usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:46

  Expected error:
      <*errors.StatusError | 0x4c2081da900>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {SelfLink: "", ResourceVersion: ""},
              Status: "200",
              Message: "",
              Reason: "",
              Details: nil,
              Code: 0,
          },
      }
      
  not to have occurred

  /usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:93
------------------------------
SSSSSSSSSSSSSSSSSSSSS

Summarizing 1 Failure:

[Fail] Cluster level logging using Elasticsearch [It] should check that logs from pods on all nodes are ingested into Elasticsearch 
/usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:93

Ran 1 of 28 Specs in 0.334 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 27 Skipped F0302 17:33:14.919906    8431 driver.go:87] At least one test failed
```
If I curl against the URL directly everything seems fine:
```
$ curl -k -u admin:Drt3KdRGnoQL7TQM  https://104.197.8.236/api/v1beta1/proxy/services/elasticsearch-logging/?namespace=default
{
  "status" : 200,
  "name" : "Gorgeous George",
  "cluster_name" : "kubernetes_logging",
  "version" : {
    "number" : "1.4.4",
    "build_hash" : "c88f77ffc81301dfa9dfd81ca2232f09588bd512",
    "build_timestamp" : "2015-02-19T13:05:36Z",
    "build_snapshot" : false,
    "lucene_version" : "4.10.3"
  },
  "tagline" : "You Know, for Search"
}
```

I assume I am either doing something wrong with the client library... or there is a bug somewhere in how the client is set up?

**Comment by roberthbailey** [comment]
I was going to do this same trick to get rid of the host port in the replication controller test (#4333). I'll give it a try today and see if I run into the same problem.

**Comment by satnam6502** [comment]
Thanks. I am 99% I simply don't understand something about how our client library works and one glance by @lavalamp will neutralize this issue. At some point when I learn enough about the client library I may even be driven to writing some documentation and examples :-)

**Comment by bgrant0607** [comment]
What happens if you leave off the namespace?

**Comment by bgrant0607** [comment]
Also, what status code was returned?

**Comment by lavalamp** [comment]
I'm not sure why your example is not working (maybe the Namespace() call is messing it up?), but there is a working example here: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/networking.go#L142

**Comment by satnam6502** [comment]
Removing the namespace does not seem to make a difference:
```
I0303 11:43:17.401401   26095 request.go:508] finalURL: https://104.197.8.236/api/v1beta1/proxy/services/elasticsearch-logging/
INFO: body: []

• Failure [0.343 seconds]
Cluster level logging using Elasticsearch
/usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:47
  should check that logs from pods on all nodes are ingested into Elasticsearch [It]
  /usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:46

  Expected error:
      <*errors.StatusError | 0x4c208032f00>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {SelfLink: "", ResourceVersion: ""},
              Status: "200",
              Message: "",
              Reason: "",
              Details: nil,
              Code: 0,
          },
      }
      
  not to have occurred

  /usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:94
------------------------------
SSSSSSSSS

Summarizing 1 Failure:

[Fail] Cluster level logging using Elasticsearch [It] should check that logs from pods on all nodes are ingested into Elasticsearch 
/usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:94

Ran 1 of 28 Specs in 0.343 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 27 Skipped F0303 11:43:17.445939   26095 driver.go:87] At least one test failed
$ curl -k -u admin:password https://104.197.8.236/api/v1beta1/proxy/services/elasticsearch-logging/
{
  "status" : 200,
  "name" : "Gorgeous George",
  "cluster_name" : "kubernetes_logging",
  "version" : {
    "number" : "1.4.4",
    "build_hash" : "c88f77ffc81301dfa9dfd81ca2232f09588bd512",
    "build_timestamp" : "2015-02-19T13:05:36Z",
    "build_snapshot" : false,
    "lucene_version" : "4.10.3"
  },
  "tagline" : "You Know, for Search"
}
```

**Comment by satnam6502** [comment]
@lavalamp changing to:
```
func getURL(c *client.Client, url string) ([]byte, error) {
	body, err := c.Get().Prefix("proxy").Resource("services").Name("elasticsearch-logging").Suffix(url).Do().Raw()
	if err != nil {
		return nil, err
	}
	return body, nil
}
```

does not seem to improve things:

```
I0303 11:49:56.359908   27067 request.go:508] finalURL: https://104.197.8.236/api/v1beta1/proxy/services/elasticsearch-logging
INFO: body: []

• Failure [0.332 seconds]
Cluster level logging using Elasticsearch
/usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:47
  should check that logs from pods on all nodes are ingested into Elasticsearch [It]
  /usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:46

  Expected error:
      <*errors.StatusError | 0x4c208174500>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {SelfLink: "", ResourceVersion: ""},
              Status: "200",
              Message: "",
              Reason: "",
              Details: nil,
              Code: 0,
          },
      }
      
  not to have occurred

  /usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:93
------------------------------
SSSSSS

Summarizing 1 Failure:

[Fail] Cluster level logging using Elasticsearch [It] should check that logs from pods on all nodes are ingested into Elasticsearch 
/usr/local/google/home/satnam/gocode/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:93

Ran 1 of 28 Specs in 0.333 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 27 Skipped F0303 11:49:56.404055   27067 driver.go:87] At least one test failed
```

**Comment by lavalamp** [comment]
Ohhh I think I see what's going on... the clue is that you are getting ```"status" : 200,``` in the body and that is causing it think it successfully parsed as an api.Status, which it's then returning to you as an error. Wow, that is insidious. I have to run to a meeting but I will think about what to do for this.

**Comment by satnam6502** [comment]
I am debugging this now. I can see that the request gets the correct response from the `Do` but then things go bad I think in `transformResponse` in `request.go`

**Comment by satnam6502** [comment]
I guess the root of the problem is this for a regular requesting expecting "200" vs. an API call expecting "Success"
```
	// If the server gave us a status back, look at what it was.
	if isStatusResponse && status.Status != api.StatusSuccess {
		// "Working" requests need to be handled specially.
		// "Failed" requests are clearly just an error and it makes sense to return them as such.
		return nil, false, errors.FromObject(&status)
	}
```

**Comment by satnam6502** [comment]
I've proposed #5002 as a fix.

**Comment by satnam6502** [comment]
Fixed in #5002
